logstash-core/lib/logstash/codecs/base.rb

```diff
@@ -28,10 +29,56 @@ def decode(data)
   alias_method :<<, :decode
 
   public
+  # DEPRECATED: Prefer defining encode_sync or multi_encode
   def encode(event)
-    raise "#{self.class}#encode must be overidden"
+    @on_event.call(event, encode_sync(event))
   end # def encode
 
+  # This is a preferred way (along with multiencode) to define a codec's encode function.
+  # This default implementation should really never be invoked by an output plugin
+  # this implementation merely provided for completeness. Plugins should prefer to simply use the
+  # concurrent_multi_receive API which always invokes multi_encode.
+  def encode_sync(event)
+    event, data = multi_encode([event]).first
+    data
+  end
+
+  public
+  # Relies on the codec being synchronous (which they all are!)
+  # We need a better long term design here, but this is an improvement
+  # over the current API for shared plugins
+  # It is best if the codec implements this directly
+  def multi_encode(events)
+    batch = if @has_encode_sync              
```

Comment:
This long `if..else..end` should be moved to a method by itself to improve readability.